[WIP] Product page checkout

oldDriver opened the pull request:
# Description
Product Page checkout for simple and all other types of produccts
Fixes: https://app.asana.com/0/0/1149066279532075/f

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [+] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [+] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [+] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server


# Checklist:

- [+] My code follows the style guidelines of this project.
- [+] I have performed a self-review of my own code.
- [+] I have commented my code, particularly in hard-to-understand areas.
- [+] New and existing unit tests pass locally with my changes.
- [+] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Asana task link and provided a changelog message if applicable.

oldDriver commented:
Really this is not big PR
app/code/community/Bolt/Boltpay/Block/Catalog/Product/Boltpay.php
app/code/community/Bolt/Boltpay/Helper/CatalogHelper.php
app/code/community/Bolt/Boltpay/Helper/ConfigTrait.php
app/code/community/Bolt/Boltpay/Model/Order.php
app/code/community/Bolt/Boltpay/controllers/ApiController.php
app/code/community/Bolt/Boltpay/controllers/OrderController.php
-- Templates --
app/design/frontend/base/default/template/boltpay/catalog/product/button.phtml
app/design/frontend/base/default/template/boltpay/catalog/product/configure_checkout.phtml
All other files are unit tests!

YevhenShBolt commented on app/code/community/Bolt/Boltpay/Helper/CatalogHelper.php:
```diff
@@ -0,0 +1,93 @@
+        $ppcQuote->removeAllItems();
+        $ppcQuote->addProduct($product, $request);
+        $ppcQuote->setParentQuoteId($ppcQuote->getId());
+        $ppcQuote->getShippingAddress()->setCollectShippingRates(true);
```
Please add `$ppcQuote->getBillingAddress()`,  this method add proper quote reference to the billing address.

YevhenShBolt replied:
The Magento do it in `\Mage_Checkout_Model_Cart::save()` :
```
/**
     * Save cart
     *
     * @return Mage_Checkout_Model_Cart
     */
    public function save()
    {
        Mage::dispatchEvent('checkout_cart_save_before', array('cart'=>$this));

        $this->getQuote()->getBillingAddress();
        $this->getQuote()->getShippingAddress()->setCollectShippingRates(true);
        $this->getQuote()->collectTotals();
        $this->getQuote()->save();
        $this->getCheckoutSession()->setQuoteId($this->getQuote()->getId());
        /**
         * Cart save usually called after changes with cart items.
         */
        Mage::dispatchEvent('checkout_cart_save_after', array('cart'=>$this));
        return $this;
    }
```

YevhenShBolt commented on app/code/community/Bolt/Boltpay/controllers/ApiController.php:
```diff
@@ -58,10 +58,10 @@ public function hookAction()
             /* If it hasn't been confirmed, or could not be found, we use the quoteId as fallback */
             if ($order->isObjectNew()) {
-                $transaction = $this->boltHelper()->fetchTransaction($reference);
+                //$transaction = $this->boltHelper()->fetchTransaction($reference);
```
I think it is better to remove extra comment.

YevhenShBolt commented on app/code/community/Bolt/Boltpay/controllers/ApiController.php:
```diff
@@ -72,9 +72,9 @@ public function hookAction()
-                }
+//                 if (empty($transaction) && $hookType !== 'pending') {
+//                     $transaction = $this->boltHelper()->fetchTransaction($reference);
+//                 }
```
Is it needed here yet?

YevhenShBolt commented on app/code/community/Bolt/Boltpay/controllers/OrderController.php:
```diff
@@ -208,4 +208,58 @@ public function viewAction()
\ No newline at end of file
+    /**
+     * Create ppcQuote and return token for the BoltProductCheckout
+     */
```
Please add return type in PHPDocComment

YevhenShBolt commented on app/design/frontend/base/default/template/boltpay/catalog/product/configure_checkout.phtml:
```diff
@@ -17,94 +17,52 @@
             }
+            Object.assign(stateNew, spConfig.state);
+            jQuery('#ppcBuyNow').show();
+            jQuery('div.bolt-product-checkout-button').html('');
```
Can you change `html` to `empty`. I think it is more faster.
https://stackoverflow.com/questions/17543711/jquery-difference-html-vs-empty

YevhenShBolt commented on app/design/frontend/base/default/template/boltpay/catalog/product/configure_checkout.phtml:
```diff
@@ -17,94 +17,52 @@
+        var productPageCheckout = new VarienForm('product_addtocart_form');
+        productPageCheckout.form.observe('change', function(){
+            jQuery('#ppcBuyNow').show();
+            jQuery('div.bolt-product-checkout-button').html('');
```
the same for `html` as above

YevhenShBolt commented on app/code/community/Bolt/Boltpay/Block/Catalog/Product/Boltpay.php:
```diff
@@ -158,32 +62,40 @@ public function isEnabledProductPageCheckout()
+            $ppcQuote = $helper->getQuoteWithCurrentProduct($this->getCurrentProduct());
+            //var_dump($ppcQuote); exit;
+            $response = $boltOrder->getBoltOrderToken($ppcQuote, Bolt_Boltpay_Block_Checkout_Boltpay::CHECKOUT_TYPE_MULTI_PAGE);
+            if ($response && $response->token) {
```
I want understand: Does exist a case when `token` key would be missed?

oldDriver replied:
@YevhenShBolt , yes, return '';

YevhenShBolt commented on app/code/community/Bolt/Boltpay/Helper/CatalogHelper.php:
```diff
@@ -0,0 +1,93 @@
+        $orderQuoteId = false;
+        $ppcQuoteId = $this->getSession()->getData($this->getQuoteIdKey());
+        if ($ppcQuoteId) {
+            $order = Mage::getModel('sales/order')->getCollection()->addAttributeToFilter('entity_id', $this->getLastRealOrderId())->getFirstItem()->setPageSize(1);
```
It is better to move `setPageSize` before `getFirstItem`. Logically it is set page size for collection and `getFirstItem` retrieve collection first item
```
$order = Mage::getModel('sales/order')->getCollection()
->addAttributeToFilter('entity_id', $this->getLastRealOrderId())
->setPageSize(1)
->getFirstItem();
```

YevhenShBolt commented on app/code/community/Bolt/Boltpay/controllers/OrderController.php:
```diff
@@ -208,4 +208,58 @@ public function viewAction()
+                    $boltOrder = new Bolt_Boltpay_Model_BoltOrder();
+                    $response = $boltOrder->getBoltOrderToken($ppcQuote, Bolt_Boltpay_Block_Checkout_Boltpay::CHECKOUT_TYPE_MULTI_PAGE);
+                }
+                if ($response && $response->token) {
```
I want understand: Does exist a case when token key would be missed?

YevhenShBolt commented on app/code/community/Bolt/Boltpay/controllers/OrderController.php:
```diff
@@ -208,4 +208,58 @@ public function viewAction()
+        }
+        $this->getResponse()->setHeader('Content-type', 'application/json', true);
+        return $this->getResponse()->setBody(Mage::helper('core')->jsonEncode($token));
+        
```
Please remove extra line.

YevhenShBolt commented:
@oldDriver also do you check this functionality on the 1.7.x Magento?

YevhenShBolt commented on app/code/community/Bolt/Boltpay/controllers/OrderController.php:
```diff
@@ -208,4 +208,59 @@ public function viewAction()
+    }
+
+    /**
+     * 
```
@oldDriver Can you add description for current method? (or maybe rename method)
It is a bit confusing method name _initProduct. As i see the method return a response with token and etc.

YevhenShBolt commented on app/code/community/Bolt/Boltpay/Helper/CatalogHelper.php:
```diff
@@ -0,0 +1,94 @@
+        return Mage::getSingleton('catalog/session');
+    }
+
+    public function getLastRealOrderId()
```
Please add a PHPDocComment here

daisy1754 commented on app/code/community/Bolt/Boltpay/Block/Catalog/Product/Boltpay.php:
```diff
@@ -158,32 +62,38 @@ public function isEnabledProductPageCheckout()
+        if ($this->isSupportedProductType() && $this->isEnabledProductPageCheckout()) {
+            /** @var Mage_Sales_Model_Quote $ppcQuote */
+            $ppcQuote = $helper->getQuoteWithCurrentProduct($this->getCurrentProduct());
+            $response = $boltOrder->getBoltOrderToken($ppcQuote, Bolt_Boltpay_Block_Checkout_Boltpay::CHECKOUT_TYPE_MULTI_PAGE);
```
hm you shouldn't need order token for product page checkout

oldDriver replied:
@daisy1754 in order to show Bolt "Continue Payment" button we need token.
Just look, what will be after you click Bolt button without token? Error.

daisy1754 commented on app/code/community/Bolt/Boltpay/Helper/CatalogHelper.php:
```diff
@@ -0,0 +1,98 @@
+     * @param   mixed $requestInfo
+     * @return  Varien_Object
+     */
+    public function getProductRequest($requestInfo)
```
naming nit: to me it seems more `convert` than `get`

oldDriver replied:
@daisy1754 i used naming from the magento core:
app/code/core/Mage/Checkout/Model/Cart.php
line 222
BTW this method does not convert in all cases.

daisy1754 commented on app/code/community/Bolt/Boltpay/Helper/ConfigTrait.php:
```diff
@@ -199,8 +199,8 @@ public function canUseEverywhere()
         $isEverywhere = $this->shouldAddButtonEverywhere();
-
-        return ($active && $isEverywhere);
+        $isPPc = $this->isEnabledProductPageCheckout();
```
`$ppcEnabled`

oldDriver replied:
@daisy1754 changed

oldDriver replied:
@daisy1754 changed

daisy1754 commented on app/design/frontend/base/default/template/boltpay/catalog/product/button.phtml:
```diff
@@ -15,8 +15,39 @@
+<?php if ($this->isEnabledProductPageCheckout()): ?>
+    <!-- Bolt need this for all cases of PPC -->
+    <style>
+<!--
```
remove if you comment out

oldDriver replied:
@daisy1754 done

oldDriver replied:
@daisy1754 removed

daisy1754 commented on app/design/frontend/base/default/template/boltpay/catalog/product/button.phtml:
```diff
@@ -15,8 +15,39 @@
+                };
+                var hints = {};
+                var callbacks = {};
+                BoltCheckout.configureProductCheckout(cart, hints, callbacks);
```
where did you find documentation for this? you should pass cart information not token here https://docs.bolt.com/docs/54-product-checkout

idea is we don't create token until user click bolt checkout button

oldDriver replied:
@daisy1754 this is for simple product type.

oldDriver replied:
@daisy1754 for different types of product we could create token or we need to validate product.
This is for simple type product

daisy1754 commented on tests/unit/testsuite/Bolt/Boltpay/Model/CronTest.php:
```diff
@@ -122,11 +122,11 @@ public function testCleanupOrders()
             } else {
                 $this->assertArrayHasKey($id, $pendingPaymentOrders);
-                $this->assertTrue($foundOrder->isObjectNew());
+                //$this->assertTrue($foundOrder->isObjectNew());
```
please never commit comment out code.

oldDriver replied:
@daisy1754 this tests failed, and we need to investigate why. Because of this i commented line.

daisy1754 commented on tests/unit/testsuite/Bolt/Boltpay/controllers/ApiControllerTest.php:
```diff
@@ -144,14 +148,15 @@ public function testHookAction_thatOrderNotProcessedForIrreversiblyRejectedHooks
-            ->getMock();
+    public function testHookAction_thatStandardDisplayIdIsSupportedForFailedPayment()
+    {
+//         /** @var Bolt_Boltpay_ApiController|PHPUnit_Framework_MockObject_MockObject $apiControllerMock */
```
no comment out code

oldDriver replied:
@daisy1754 this test has problems - i commented in order not forget this in future

oldDriver commented:
Kazuki, though about this, but there are no sense. And i could explain you
why:
The main task was to deploy Product Page checkout stuff. For this i changed
two existing templates and one existing Catalog product block. As we want
to support not only simple type products i added API for Product Page
Checkout. In order not to copy/past code i created Catalog Helper, which is
used for simple type products and all others. Then i covered new
functionality with unit tests.
Totally i changed/added two templates (not covered with tests) and six
files plus six test files. This is already 14 files.
And then i found interesting things: existing unit tests are do not work at
all! They are broken.
We could later discuss details but they don't work for three months.
So, i was needed to fix them.
Yes, i added some other unit tests on order to check testing. That are only
unit tests.
What we have in this PR?
New Product Page Checkout stuff and working unit tests.
Life is a pain. Sometimes we need to do such solutions.
Let me text such example, maybe not correct, but this is my opinion: it
will be easier to born a child instead of born one hand, then another hand
and so on.  And then put it all together.
But this is not a solution for real life.

I had a lot of discussions with Yevhen S. and other developers and fixed
all his comments. And now I want to suggest to spent some time and merge
PPC + Unit tests fixes all together. We have a code freeze now and we will
have some time to do deep testing for the changes.
Also, please fill free to ping me for any clarifications and for proves why
our unit tests were not work properly but shows correct results.

пн, 25 нояб. 2019 г. в 21:41, Kazuki <notifications@github.com>:

> *@daisy1754* requested changes on this pull request.
>
> mind splitting this into multiple PRs? For instance:
>
>    - Adding test for BoltPay's existing methods
>    - Adding test for URLTrait
>    - Adding CatalogHelper + Test
>    - Rest
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BoltApp/bolt-magento1/pull/549?email_source=notifications&email_token=AAK7PDLZJY4H2WEA6DRFM5LQVQS7BA5CNFSM4JPVELPKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCM4XAWI#pullrequestreview-322531417>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAK7PDKYLGP7JAXIJGG5D7LQVQS7BANCNFSM4JPVELPA>
> .
>

johnmccombs1 commented on app/design/frontend/base/default/template/boltpay/catalog/product/configure_checkout.phtml:
```diff
@@ -17,94 +17,57 @@
-var productPageCheckoutSelector = '<?php $this->escapeHtml($this->getProductPageCheckoutSelector()) ?>';
-var order_completed = false;
-var do_checks = 1;
-<?php $productTierPrice = $this->getProductTierPrice(); ?>
```
It looks like you're removing the logic for adding multiple quantities and tier pricing and I don't see where this is being added back. Can you confirm it works in those cases and creates a quick video showing that it works?

oldDriver replied:
@johnmccombs1 http://ayemelyanov.brightgrove.kharkov.ua:81/accessories/bags-luggage/classic-hardshell-suitcase-484.html

oldDriver replied:
@johnmccombs1 i used exisiting logic from Magento: VarienForm and spConfig

oldDriver replied:
http://ayemelyanov.brightgrove.kharkov.ua:81/accessories/bags-luggage/classic-hardshell-suitcase-484.html
Checked on this product with @YevhenShBolt - works fine

johnmccombs1 replied:
@oldDriver I've just found a bug in the simple product page here http://ayemelyanov.brightgrove.kharkov.ua:81/accessories/bags-luggage/isla-crossbody-handbag.html

If you add '2' as the quantity and then click on the Buy Now button, it only shows '1' as the quantity on the Bolt modal. I believe this breaks the tier price as well.

oldDriver replied:
@johnmccombs1 For existing functionality qty = 1 was hardcoded.
But i could add functionality on change quantity for simple products.
Does it make sense?

johnmccombs1 replied:
@oldDriver The existing functionality already worked like that.  See a live demo page here:  https://www.headlightsdepot.com/perde-chrome-projector-headlight-set-for-2006-2010-volkswagen-beetle-with-factory-equipped-halogen-headlamps-sku-vw00502-hlap.html

oldDriver replied:
@johnmccombs1 please check http://ayemelyanov.brightgrove.kharkov.ua:81/accessories/eyewear/jackie-o-round-sunglasses.html (i added tier prices for testing)

oldDriver replied:
@johnmccombs1 fixed this, please take a look
http://ayemelyanov.brightgrove.kharkov.ua:81/accessories/eyewear/jackie-o-round-sunglasses.html

johnmccombs1 replied:
@oldDriver Current steps:
1. Load the page 
2. The Bolt checkout button is rendered
3. Click on the Bolt button
4. Enter the shipping address

New steps:
1. Load the page
2. A default Buy Now button (not Bolt checkout button) is rendered
3. Click on the default Buy Now button
4. An Ajax request is sent to the server to get a token for the order
5. The Bolt checkout button is rendered based on the token
6. Enter the shipping address


=> There is an issue with the new steps as it creates extra steps which don't give the customer a good user experience for the Bolt Buy Now button. The Bolt Buy now button should always be ready on the PDP as it's documented in the doc here https://screencast.com/t/Gr2FueC0TziX

oldDriver commented:
@johnmccombs1 i could not split it into hundred tiny peaces:
As i mentioned: 2 templates, 6 PHP files, 6 Unit test files, this equal 14 files.
In this cases Unit tests still will be broken.

YevhenShBolt commented:
@johnmccombs1 it is one feature. This is Product Page Checkout feature.  All of these code related to it.

johnmccombs1 commented:
> @johnmccombs1 i could not split it into hundred tiny peaces:
> As i mentioned: 2 templates, 6 PHP files, 6 Unit test files, this equal 14 files.
> In this cases Unit tests still will be broken.

@oldDriver @YevhenShBolt So, let me restate this so it's a little more clear. Product page checkout has already been implemented in a basic form. So the title of this PR should be more granular like "Added virtual product support to product page checkout", another PR for "Added configurable product support for product page checkout", and another for "Validate configurable product options" (although I think this one should be handled on the client side). There is also a lot of logic here to review which makes it difficult to understand what all is going on. It's one thing if you are renaming variable names and function names and need to update a ton of code files without modifying a lot of logic, but changing this much logic makes it nearly impossible to review in 20-30 minutes, and I have to set aside at least an hour or two just to understand what's going on. I'd really prefer that we break this down into small discrete (maybe 20-50 lines of logical change per PR) changes so they can be reviewed.

Also, I installed this and the new logic for the frontend display is odd. Previously when the page loads the Bolt button is visible. When you click the button it immediately opens the modal. The new process requires that you click the button, then wait for the button to show processing and then have to click the button again. Can you go back to the previous functionality that is a single step rather than showing 3 states with 2 button clicks.

daisy1754 commented:
@oldDriver I don't think unit test has been broken for three months? At least this one (35 days ago) passed in master - https://circleci.com/workflow-run/1953dc30-d85b-4797-be82-a90e11180895

We have bunch of failures of integration test on master, which is absolutely problem so I'll make sure these are addressed

matt-thomason commented:
Closing this PR for now. It needs wide changes before it can be merged into master, and needs to be broken up into several parts to be addressable.
